src/curve.rs

```diff
@@ -0,0 +1,326 @@
+    /// multiplication by doubling.
+    // TODO: implement Daira's algorithm from https://github.com/zcash/zcash/issues/3924
+    pub fn scalar_mult(
+        builder: &mut GadgetBuilder<F>,
+        point: &EdwardsPointExpression<F, C>,
+        scalar: &Expression<F>
+    ) -> EdwardsPointExpression<F, C> {
+        let scalar_binary = builder.split_allowing_ambiguity(&scalar);
+
+        let mut sum = Self::identity();
+        let mut current = point;
+        for bit in scalar_binary.bits {
+            let boolean_product= &Self::boolean_mult(builder, current, &bit);
+            sum = Self::add(builder, &sum,boolean_product);
+            let current = &Self::double(builder, current);
+        }
+        sum
+    }
+
+    /// Given a boolean element, return the given element if element is on, otherwise
+    /// return the identity.
+    fn boolean_mult(
+        builder: &mut GadgetBuilder<F>,
+        point: &EdwardsPointExpression<F, C>,
+        boolean: &BooleanExpression<F>,
+    ) -> EdwardsPointExpression<F, C> {
+        let x = builder.selection(boolean, &point.x, &Expression::zero());
+        let y = builder.selection(boolean, &point.y, &Expression::one());
+        EdwardsPointExpression::new_unsafe(x, y)
+    }
+
+    /// Identity element for twisted Edwards Curve
+    pub fn identity() -> EdwardsPointExpression<F, C> {
+        EdwardsPointExpression::new_unsafe(Expression::zero(), Expression::one())
+    }
+
+    /// Takes two elements as coordinates, checks that they're on the curve without adding
+    /// constraints, and then returns an EdwardsPointExpression
+    pub fn from_elements(x: Element<F>, y: Element<F>) -> EdwardsPointExpression<F, C> {
+        assert!(C::a() * &x * &x + &y * &y == Element::one() + C::d() * &x * &x * &y * &y,
+                "Point must be contained on the curve.");
+        EdwardsPointExpression::new_unsafe(Expression::from(x), Expression::from(y))
+    }
+
+    /// Takes two expressions as coordinates, adds constraints verifying that the coordinates
+    /// are contained on the specified curve, and then returns an EdwardsPointExpression
+    pub fn from_expressions(builder: &mut GadgetBuilder<F>, x: Expression<F>, y: Expression<F>) -> EdwardsPointExpression<F, C> {
+        let x_squared = builder.product(&x, &x);
+        let y_squared = builder.product(&y, &y);
+        let x_squared_y_squared = builder.product(&x_squared, &y_squared);
+        builder.assert_equal(&(&x_squared * C::a() + &y_squared),
+                             &(&x_squared_y_squared * C::d() + Expression::one()));
+        EdwardsPointExpression::new_unsafe(x, y)
+    }
+
+    /// Takes two expressions as coordinates, does not perform a check or add constraints
+    /// to check that the coordinates are on the specified curve, and then returns an
+    /// EdwardsPointExpression
+    pub fn new_unsafe(x: Expression<F>, y: Expression<F>) -> EdwardsPointExpression<F, C> {
```

Comment:
Maybe rename to `from_elements_unsafe`? Or could rename `from_elements` to `new`.

Reply:
Should we do from_expressions_unsafe?